server/src/main/java/org/opensearch/action/search/TransportSearchAction.java

```diff
@@ -433,11 +441,16 @@ private void executeRequest(
         }
         SearchRequestOperationsListener.CompositeListener requestOperationsListeners = searchRequestOperationsCompositeListenerFactory
             .buildCompositeListener(originalSearchRequest, logger);
-        SearchRequestContext searchRequestContext = new SearchRequestContext(requestOperationsListeners, originalSearchRequest);
+        SearchRequestContext searchRequestContext = new SearchRequestContext(requestOperationsListeners, originalSearchRequest, tracer);
```

Comment:
@dzane17 I believe the implementation should not modify `SearchRequestContext` in any form, please keep in mind that tracing should be as zero cost as possible and not leak into throughout the codebase. The `SearchRequestOperationsListener` already has the context demarcation we need, I would suggest:
 - create new class `TracebaleSearchRequestOperationsListener`: this class would keep the reference to current parent and start / end new spans on onXxxStart  / onXxxEnd / onXxxFailure callbacks
- we add this class to `SearchRequestOperationsListener.CompositeListener` only if tracing is enabled (check `Tracer` API, it has methods for it)
- the  `SearchRequestContext`  stays clean from any knowledge related to tacing

Reply:
We had a similar discussion on previous `SearchRequestOperationsListener` subscribers. Right now I am creating a single `SearchRequestCoordinatorTrace` **per node**, however with `TraceableSearchRequestOperationsListener` we will need to create it **per request**.

Based on previous instrumentation PRs and discussion I thought we do not need to check if tracing FeatureFlag is enabled as that is handled by the tracer itself (Noop tracer). 

I understand keeping `SearchRequestContext` but it comes at a cost.

Reply:
> Based on previous instrumentation PRs and discussion I thought we do not need to check if tracing FeatureFlag is enabled as that is handled by the tracer itself (Noop tracer).

The tracer could be enabled / disabled dynamically, I was mentioning the `Tracer` API which has `isRecording` method, if we do not record, we should not create instrumentation

> TraceableSearchRequestOperationsListener we will need to create it per request

The trace is per request, right?

Reply:
We need the tracer and spans to set up the span scope
```
try (SpanScope spanScope = tracer.withSpanInScope(searchRequestContext.getPhaseSpan())) {
    phase.recordAndRun();
}
```
If we store everything in `TraceableSearchRequestOperationsListener` then it is not accessible in AbstractSearchAsyncAction.

Reply:
> If we store everything in `TraceableSearchRequestOperationsListener` then it is not accessible in AbstractSearchAsyncAction.

Why do you need to access it? The span will be started by the traceable search listener in `onPhaseStart` and finished in `onPhaseEnd`

Reply:
See in this example, we do not have `phaseSpan` to set up SpanScope
```
onPhaseStart(phase, searchRequestContext);
try (SpanScope spanScope = tracer.withSpanInScope(phaseSpan)) {
    phase.recordAndRun();
}
```

`phaseSpan` is stored in `TraceableSearchRequestOperationsListener` inside `SearchRequestOperationsListener.CompositeListener`

Reply:
**phaseSpan** is coming from current thread:  `onPhaseStart(phase);` starts the scoped span for phase, the `phase.recordAndRun();` inherit it, `onPhaseEnd/Failure(phase);` the scoped span is closed.

Reply:
Got it! I've replaced `SearchRequestCoordinatorTrace` with `TraceableSearchRequestOperationsListener`